ec2-spot-sagemaker-managed-spot-training/sagemaker-built-in-object-detection/tools/birdsOnEdge.py

```diff
@@ -0,0 +1,225 @@
+    with zipfile.ZipFile("/tmp/boto_3_dist.zip", "r") as zip_ref:
+        zip_ref.extractall(boto_dir)
+client.publish(topic=iot_topic, payload="Adding boto to path...")
+sys.path.append(boto_dir)
+
+import boto3
+
+client.publish(topic=iot_topic, payload="Completed import of boto3")
+
+
+class LocalDisplay(Thread):
+    """Class for facilitating the local display of inference results
+    (as images). The class is designed to run on its own thread. In
+    particular the class dumps the inference results into a FIFO
+    located in the tmp directory (which lambda has access to). The
+    results can be rendered using mplayer by typing:
+    mplayer -demuxer lavf -lavfdopts format=mjpeg:probesize=32 /tmp/results.mjpeg
+    """
+
+    def __init__(self, resolution):
+        """resolution - Desired resolution of the project stream"""
+        # Initialize the base class, so that the object can run on its own
+        # thread.
+        super(LocalDisplay, self).__init__()
+        # List of valid resolutions
+        RESOLUTION = {"1080p": (1920, 1080), "720p": (1280, 720), "480p": (858, 480)}
+        if resolution not in RESOLUTION:
+            raise Exception("Invalid resolution")
+        self.resolution = RESOLUTION[resolution]
+        # Initialize the default image to be a white canvas. Clients
+        # will update the image when ready.
+        self.frame = cv2.imencode(".jpg", 255 * np.ones([640, 480, 3]))[1]
+        self.stop_request = Event()
+
+    def run(self):
+        """Overridden method that continually dumps images to the desired
+        FIFO file.
+        """
+        # Path to the FIFO file. The lambda only has permissions to the tmp
+        # directory. Pointing to a FIFO file in another directory
+        # will cause the lambda to crash.
+        result_path = "/tmp/results.mjpeg"
+        # Create the FIFO file if it doesn't exist.
+        if not os.path.exists(result_path):
+            os.mkfifo(result_path)
+        # This call will block until a consumer is available
+        with open(result_path, "w") as fifo_file:
+            while not self.stop_request.isSet():
+                try:
+                    # Write the data to the FIFO file. This call will block
+                    # meaning the code will come to a halt here until a consumer
+                    # is available.
+                    fifo_file.write(self.frame.tobytes())
+                except IOError:
+                    continue
+
+    def set_frame_data(self, frame):
+        """Method updates the image data. This currently encodes the
+        numpy array to jpg but can be modified to support other encodings.
+        frame - Numpy array containing the image data of the next frame
+                in the project stream.
+        """
+        ret, jpeg = cv2.imencode(".jpg", cv2.resize(frame, self.resolution))
+        if not ret:
+            raise Exception("Failed to set frame data")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.